test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+        t.data = lhs.data + rhs.data;
+        return t;
+    }
+    operator +=(ref lhs: Tensor(?d), const ref rhs: Tensor(d)) {
+        if lhs.domain.size == rhs.domain.size {
+            lhs.data += rhs.data;
+        } else if lhs.domain.size == 0 && rhs.domain.size != 0 {
+            lhs.reshapeDomain(rhs.domain);
+            lhs.data = rhs.data;
+        } else if lhs.domain.size != 0 && rhs.domain.size == 0 {
+            // do nothing
+        } else {
+            err("Cannot add tensors of different sizes. += ", lhs.domain.size, " != ", rhs.domain.size,"  [",lhs.shape," += ",rhs.shape,"]");
+        }
+    }
+    operator +=(ref lhs: Tensor(?rank,?eltType), rhs) where (isArray(rhs) && rhs.rank == rank) || rhs.type == eltType {
+        lhs.data += rhs;
+    }
+    operator +=(ref lhs: Tensor(?rank,?eltType), rhs) where rhs.type:string == "promoted expression" || rhs.type:string == "iterator" {
+        lhs.data += rhs;
+    }
+    operator -(lhs: Tensor(?rank,?eltType), rhs: Tensor(rank,eltType)) {
+        var t = new Tensor(rank=rank,eltType=eltType);
+        t.reshapeDomain(lhs._domain);
+        t.data = lhs.data - rhs.data;
+        return t;
+    }
+    operator -=(ref lhs: Tensor(?d), const ref rhs: Tensor(d)) {
+        lhs.data -= rhs.data;
+    }
+    operator -=(ref lhs: Tensor(?rank,?eltType), rhs) where (isArray(rhs) && rhs.rank == rank) || rhs.type == eltType {
+        lhs.data -= rhs;
+    }
+    operator *(c: ?eltType, rhs: Tensor(?rank,eltType)) {
+        var t = new Tensor(rank=rank,eltType=eltType);
+        t.reshapeDomain(rhs._domain);
+        t.data = c * rhs.data;
+        return t;
+    }
+    operator *(lhs: Tensor(?rank,?eltType), c: eltType) {
+        var t = new Tensor(rank=rank,eltType=eltType);
+        t.reshapeDomain(lhs._domain);
+        t.data = lhs.data * c;
+        return t;
+    }
+    operator *(lhs: Tensor(?rank,?eltType), rhs: Tensor(rank,eltType)) {
+        // Hermitian product, not composition
+        var t = new Tensor(rank=rank,eltType=eltType);
+        t.reshapeDomain(lhs._domain);
+        t.data = lhs.data * rhs.data;
+        return t;
+    }
+    operator *=(ref lhs: Tensor(?d), const ref rhs: Tensor(d)) {
+        lhs.data *= rhs.data;
+    }
+    operator *=(ref lhs: Tensor(?rank,?eltType), rhs) where (isArray(rhs) && rhs.rank == rank) || rhs.type == eltType {
+        lhs.data *= rhs;
+    }
+
+    // Matrix-vector multiplication
+    operator *(lhs: Tensor(2,?eltType), rhs: Tensor(1,eltType)): Tensor(1,eltType) {
+        const (m,n) = lhs.shape;
+        const (p,) = rhs.shape;
+        if n != p then
+            err("Trying to apply a matrix of shape ",lhs.shape, " to a vector of shape ", rhs.shape);
+
+        const a = lhs.data;
+        const v = rhs.data;
+        var w = new Tensor(rank=1,eltType=eltType);
+        w.reshapeDomain({0..#m});
+        forall i in 0..#m with (ref w) {
+            const row = a[i,..];
+            w[i] = + reduce (row * v);
```

Comment:
I believe rank reducing slices like `a[i,..]` are pretty inefficient currently. I think it might be more efficient to rewrite this as a doubly nested loop instead of using a reduction over the array slice:

```chapel
forall i in 0..#m with (ref w) {
  for j in 0..#n {
    w[i] += row[i, j] * v;
  }
}
```

It would be even more efficient to use a chunking algorithm here but I'm not sure you'd see much benefit with the size of matrix you're working with since they would mostly fit into cache.